src/data_conn.c

```diff
@@ -210,22 +215,20 @@ handle_data_conn_error(replica_t *r)
 	spec = r->spec;
 
 	TAILQ_FOREACH(r1, &(spec->rq), r_next)
-		if (r1 == r)
+		if (r1 == r) {
+			update_spec_status= 1;
 			break;
-
-	if (r1 == NULL) {
-		REPLICA_ERRLOG("replica %s %d not part of rqlist..\n",
-		    r->ip, r->port);
-		MTX_UNLOCK(&spec->rq_mtx);
-		return -1;
-	}
-	TAILQ_REMOVE(&spec->rq, r, r_next);
+		}
 
 	if (r->state == ZVOL_STATUS_HEALTHY)
 		spec->healthy_rcount--;
 	else if (r->state == ZVOL_STATUS_DEGRADED)
 		spec->degraded_rcount--;
-	update_volstate(r->spec);
```

Comment:
why we need these changes here? there is lock on spec, so, removing from list should be fine.